design/proposals/variable_schema_validation.md

```diff
@@ -0,0 +1,264 @@
+      resource: *vars
+```
+
+## Proposed Solution
+
+The proposed solution is to add the ability to use JSON schemas and link them to variables, for which we want validation of their content. This will allow a single variable to store multiple, related values. While it's already possible today, to store JSON structured information in a variable, this proposed feature adds two added values:
+
+- The JSON structure is validated and enforced. If the input is invalid, the user will get a proper error message and the variable update will be denied.
+- The JSON structure will not be encrypted completely, but only the sensitive information. This will allow future seach capabilities based on the non sensitive information.
+
+### User Experience
+
+Let's have a look at the following example:
+
+```yaml
+- !policy
+  id: my-policy
+  body:
+  - !variable
+    id: mysql-db-creds
+    schema: !variable /conjur/schemas/mysql-schema
+
+  - !host my-app
+  
+  - !permit
+    role: !host my-app
+    privileges: [ read,execute ]
+    resource: !variable mysql-db-creds
+```
+
+And
+
+```yaml
+- !policy
+  id: conjur/schemas
+  body:
+  - &schemas
+    - !variable mysql-schema
+  
+  - !permit
+    role: !group /conjur/all
+    privileges: [ read,execute ]
+    resource: !variable *schemas
+```
+
+In the example above, the `/conjur/schemas/mysql-schema` variable will be updated with the following value:
+
+```json
+{
+  "$id": "https://cyberark.com/mysql.schema.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "MySQL Creds",
+  "type": "object",
+  "properties": {
+    "username": {
+      "type": "string",
+      "description": "The MySQL username"
+    },
+    "password": {
+      "type": "string",
+      "description": "The MySQL password",
+      "minLength": 12,
+      "maxLength": 40,
+      "pattern": "[A-Za-z0-9]"
+    },
+    "address": {
+      "type": "string",
+      "description": "The MySQL address"
+    },
+    "port": {
+      "description": "The MySQL port number",
+      "type": "integer",
+      "minimum": 1,
+      "maximum": 65535
+    }
+  },
+  "required": ["username", "password"],
+  "additionalProperties": false,
+  "secrets": [
+      "password",
+    ]
+}
+```
+
+And the `mysql-db-creds` variable could be updated with the following value:
+
+```json
+{
+  "username": "my-user",
+  "password": "my-password-12345",
+  "address": "my-db.com",
+  "port": 3306
+}
+```
+
+### Functionality
+
+#### Input Validation
+
+Conjur will validate the variable content, using the JSON schema that is specified in the `schema` attribure of that variable.
+
+For the example above, Conjur will use the JSON schema specified in `conjur/schemas/mysql-schema` to verify the content of `mysql-db-creds`. This verification includes:
+
+- `username` and `password` are mandatory properties.
+- `address` and `port` are allowed properties.
+- `port` must contain a valid port number.
+- `password` must align to a sufficient complexity.
```

Comment:
What about allowed characters?

Reply:
Allowed and disallowed characters are supported in JSON schema. In my example, the port has to be a number between 1 and 65535.

Reply:
I agree that REGEX would be a good solution here, by it is impossible in JSON